server.js

```diff
@@ -239,11 +241,31 @@ module.exports = function(options) {
   // This will catch all 404s and redirect them to root URL
   // with preserving the pathname for client-side to handle.
   server.ext('onPreResponse', function(request, reply) {
+    var url = request.url.pathname;
+    var pathname = url.split('/')[2];
+    var query = '';
+    if (request.url.search) {
+      query = request.url.search;
+    }
     if (request.response.output && request.response.output.statusCode === 404) {
-      return reply.redirect('/?redirect=' + request.url.pathname);
+      if (pathname && paths.findIndex(function(path) { return path.pathname === '/' + pathname + '/'; }) !== -1) {
+        return reply.redirect('/' + pathname + '/' + query);
+      }
+      return reply.redirect('/' + query);
     }
     return reply.continue();
   });
+  server.ext('onRequest', function(request, reply) {
+    var query = '';
```

Comment:
I feel like there's an easier way to implement this logic, I'm just not sure how...